### DISCUSSION TITLE
How to add a fixture from pytest_generate_tests?

### DISCUSSION THREAD

**Discussion by amitw-di** [pytest-dev/pytest#12205] [Q&A] [answered]
Hi, 
I tried to add a fixture inside pytest_generate_tests, and while I've been able to add something, the fixture code is not working. 

example of my attempt:

```
   def pytest_generate_tests(self, metafunc):
        marks = metafunc.definition.iter_markers('use_myfix')
        if len(list(marks)) > 0 and myOS == 'Windows':
            metafunc.fixturenames.append('myfix')
            metafunc.parametrize('myfix', ['myfix-1', 'myfix-2', 'myfix-3'], indirect=True)
```

The test, naturally, is not aware of the "myfix" parameter, and I want to have this similar to an autouse fixture (only for the relevant test). 

I have seen the question here:https://github.com/pytest-dev/pytest/discussions/10686, but at least since this PR: https://github.com/pytest-dev/pytest/pull/11209, the answer there is not possible. Even after trying to copy the FuncFixtureInfo object, code isn't being executed. 
I also tried the workaround suggested there with a wrapper fixture, only it seems that the args added in pytest_generate_tests are not being added to the ones in request (which is quite odd). 

Any ideas?

**Comment by RonnyPfannschmidt** [comment]
alering fixturenames has no safe effect - please add a usefixtures marker to the metafunc.definition node

**Reply by amitw-di** [reply]
Thanks for the answer, I'm probably doing something wrong, since this seems to have no effect. 

```
    metafunc.definition.own_markers.append(pytest.mark.usefixtures('myfix'))
    # metafunc.definition.fixturenames.append('myfix')
    metafunc.parametrize('myfix', ['myfix-1', 'myfix-2', 'myfix-3'], indirect=True)
```

Without the commented out line, I'm getting  "function uses no fixture 'myfix' "

With the commented out line stuff run, but the fixture's code is not being invoked. 

what am I missing?

**Reply by RonnyPfannschmidt** [reply, marked as answer]
this is possibly a interaction bug - mast likely related to when/how fixturenames is set up

we need to open a issue on this (as its not something i myself can work on soonish)
